src/main/java/org/eclipse/uprotocol/streamer/UStreamer.java

```diff
@@ -0,0 +1,183 @@
+ * 
+ * The streamer provides APIs to simply add remove forwarding rules. A rule is defined as a 
+ * combination of input and output routes. The input route is the route to listen to and the output route is the route to forward the message to.
+ * The forwarding works by registering a listener on the input route and forwarding the message to the output route.
+ */
+public class UStreamer {
+
+    // List of listeners (routes) that the streamer listens to
+    private List<TransportListener> listeners;
+
+    private USubscription submgr;
+
+    // Listener to handle dynamic changes to subscriptions for anyone
+    private class SubscriptionChangeListener implements UListener {
+
+        @Override
+        public void onReceive(UMessage message) {
+            // TODO Auto-generated method stub
+            throw new UnsupportedOperationException("Unimplemented method 'onReceive'");
+        }
+
+    }
+
+    /**
+     * Constructor
+     */
+    public UStreamer(USubscription submgr) {
+        Objects.requireNonNull(submgr, "summgr cannot be null");
+        this.submgr = submgr;
+        listeners = new ArrayList<>();
+
+        // TODO: Register to be notified of any and all subscription changes
+        //submgr.registerForNotifications(NotificationsRequest.newBuilder().setTopic())
+    }
+
+
+    /**
+     * Add a forwarding rule to the streamer. 
+     * The input route is the route to listen to and the output route is the route to forward the message to.
+     * The forwarding works by registering a listener on the input route and forwarding the message to the output route.
+     * 
+     * Routing rules work on destination address for notifications, requests, and response messages however for
+     * publish messages (that do not have a destination address) we use uSubscription service to fetch remote
+     * subscriptions. 
+     * NOTE: Default routing rule (where UAuthority.name is "*") *MUST* be added at the end of the list to avoid overriding other rules.
+     * 
+     * @param in input {@code Route} that the streamer listens to
+     * @param out output {@code Route} that the streamer forwards the message to
+     * @return {@code UStatus} with UCode.OK if the forwarding rule was added successfully,
+     *        {@code UStatus} with UCode.ALREADY_EXISTS if the forwarding rule already exists,
+     *       {@code UStatus} with UCode.INVALID_ARGUMENT if the input and output routes are the same
+     */
+    public UStatus addForwardingRule(Route in, Route out) {
+        Objects.requireNonNull(in, "input cannot be null.");
+        Objects.requireNonNull(out, "output cannot be null.");
+        
+        // Cannot route to itself
+        if (in.equals(out)) {
+            return UStatus.newBuilder().setCode(UCode.INVALID_ARGUMENT).build();
+        }
+
+        // check if the rule already exists in the list
+        if (listeners.stream().anyMatch(p -> p.getInputRoute().equals(in) && p.getOutputRoute().equals(out))) {
```

Comment:
this is not very efficient. An O(n) check can be a problem. If the list is small no problem. Maybe a data structure like a Set or a Map can help with managing duplicates?

Reply:
do we need to do this optimization in the demo code though... Maybe I need to move this code out of up-java and into a separate project so that it is not accidentally used by developers!!